genson/src/main/java/com/owlike/genson/convert/CircularClassReferenceConverterFactory.java

```diff
@@ -18,15 +19,24 @@
  */
 public class CircularClassReferenceConverterFactory extends ChainedFactory {
   private final static class CircularConverter<T> extends Wrapper<Converter<T>> implements Converter<T> {
+
+    private CountDownLatch initLatch = new CountDownLatch(1);      
+
     protected CircularConverter() {
       super();
     }
 
     public void serialize(T obj, ObjectWriter writer, Context ctx) throws Exception {
+      if(wrapped == null) {
+        initLatch.await();
```

Comment:
I do see a few advantages in moving the "locking" logic to the factory.
  - The converters are supposed to have immutable state once started being used to do ser/de.
  - I like the idea of isolating the initialization of converters in the factories
  - Even if less efficient the creation process happens only at startup for each type, so the cost of doing locking or synchronization is low.
  - It is simpler.

Reply:
I agree, however it's not sufficient to fix the issue - race-condition happens between _create_ and _serialise_  invocations, _create_ uses thread-local and should not have any synchronisation issues. Non-initialized instance of CircularConverter leaks out _create_ recursive call through another converter (created via 'next().create(..)') - that causes it being used before it has finished initialising.

Reply:
Non-initialized CircularConverter gets to another thread via stack-trace that should look like:
```
CircularClassReferenceConverterFactory.create()
  CircularClassReferenceConverterFactory > _circularConverters.get().put(type, new CircularConverter())
    CircularClassReferenceConverterFactory > next().create()
      ??ConverterFactory.create()
        ??ConverterFactory > next().create()
          CircularClassReferenceConverterFactory.create()
            CircularClassReferenceConverterFactory > return _circularConverters.get().get(type)
```
??ConverterFactory.create - creates converter with non-initialzied child of CircularConverter (from thread-local), this converter gets used from another thread before 'CircularClassReferenceConverterFactory > next().create()' finishes invocation.

Reply:
I should have read the code more carefully, I missed the thread local :/
Indeed it will leak through the bean descriptor that is being constructed and that will use Genson.provideConverter on the recursive type, which in turn will put the uninitialized converter in the shared converter map :/

Interestingly I haven't been able to write a code that hits the same issue when using the lock in factory.create. Although I agree that the same problem should happen...

Reply:
I found why I couldn't reproduce it with the lock in the factory. The factory will be used for every type, even ones that don't have a circular reference. So doing the locking in there will block all the threads that want to create a converter, so they will not reach the code that reads from Genson.provideConverter. However if someone was trying to ser/de directly the circular type, then the problem could occur again.

Considering all that your solution looks great :) Good that you took the time to explain me what was wrong.